src/Psl/Type/Internal/VecType.php

```diff
@@ -87,22 +93,26 @@ public function coerce(mixed $value): iterable
     public function assert(mixed $value): array
     {
         if (! is_array($value) || !array_is_list($value)) {
-            throw AssertException::withValue($value, $this->toString(), $this->getTrace());
+            throw AssertException::withValue($value, $this->toString());
         }
 
-        /** @var Type\Type<Tv> $value_type */
-        $value_type = $this->value_type->withTrace(
-            $this->getTrace()
-                ->withFrame('vec<' . $this->value_type->toString() . '>')
-        );
-
         $result = [];
-
-        /**
-         * @var Tv $v
-         */
-        foreach ($value as $v) {
-            $result[] = $value_type->assert($v);
+        $value_type = $this->value_type;
+        $i = $v = null;
+
+        try {
+            /**
+             * @var Tv $v
+             * @var array-key $i
+             */
+            foreach ($value as $i => $v) {
+                $result[] = $value_type->assert($v);
+            }
+        } catch (AssertException $e) {
+            throw match (true) {
+                $i === null => $e,
```

Comment:
same as before, it is impossible for `$i` to be null here, but it is possible in coerce because `$value` could be an object that throws when you try to iterate over it.